metronome/checkpointing/models/src/io/iohk/metronome/checkpointing/models/CheckpointCertificate.scala

```diff
@@ -31,14 +34,65 @@ case class CheckpointCertificate(
 )
 
 object CheckpointCertificate {
+
+  /** Create a `CheckpointCertificate` from a `Block` that last had a `CheckpointCandidate`
+    * and a list of `Block.Header`s leading up to the `QuorumCertifictate` that proves the
+    * BFT agreement over the contents.
+    */
   def construct(
       block: Block,
       headers: NonEmptyList[Block.Header],
       commitQC: QuorumCertificate[CheckpointingAgreement, Phase.Commit]
-  ): Option[CheckpointCertificate] =
+  ): Option[CheckpointCertificate] = {
+    assert(block.hash == headers.head.hash)
     constructProof(block).map { case (proof, cp) =>
       CheckpointCertificate(headers, cp, proof, commitQC)
     }
+  }
+
+  /** Validate a `CheckpointCertificate` by checking that:
+    * - the chain of block headers is valid
+    * - the quorum certificate is valid
+    * - the Merkle proof of the candidate is valid
+    */
+  def validate(
+      certificate: CheckpointCertificate,
+      federation: Federation[CheckpointingAgreement.PKey]
+  )(implicit
+      signing: Signing[CheckpointingAgreement]
+  ): Either[String, Validated[Transaction.CheckpointCandidate]] = {
+    def check(
+        message: String
+    )(cond: Boolean): Either[String, Unit] =
+      Either.cond(cond, (), message)
+
+    for {
+      _ <- check(
+        "The headers do not correspond to a chain of parent-child blocks."
```

Comment:
Would it make sense to say this in positive terms (remove "do not"):
"The headers correspond to a chain of parent-child blocks." etc.

Reply:
I know what you mean, it looks like a statement that already happened, but it becomes an error message where it makes sense this way :-/ 

Scalaz used to have some syntax for validation where it was like `condtion \/ "Whatever failed."`, I'll see if I can reorder it to make it look more natural.

Reply:
Changed it to `test orError "This failed."` pattern. Does that work better?

Reply:
I guess alternatively I could use `must be` as a positive term, like `check("The Commit Q.C. must be valid.", signing.validate(commitQC))`.

Reply:
I was not aware that this becomes error message. In that case I am fine with this :)